### ISSUE TITLE
Support setting UID and GID via authctl

### ISSUE THREAD

**Issue by callanova** [canonical/authd#630]
### Is there an existing request for this feature?

- [X] I have searched the existing issues and found none that matched mine

### Describe the feature

Currently, authd creates a random UID and GID for any users / groups it creates.
For example, during first authentication, firstname.lastname@company.com creates a UID of 1021432359 and a GID of 1021432359.
We have a workflow that requires a user's UID and GID to be set to a specific value like 1001.
usermod and groupmod don't allow us to modify the UID and GID of firstname.lastname@company.com because, the user and group don't exist in /etc/passwd and /etc/group respectively.
It would be great if there was a way to modify a UID and GID for users / groups created by authd.

### Describe the ideal solution

A CLI cmd that can manage and modify /var/cache/authd/auth.db allowing us the ability to update the UID and GID for users / groups created by authd.

### Alternatives and current workarounds

Use a non-authd account.

### Environment
* broker version: please run `snap info authd-msentraid`
```
name:      authd-msentraid
summary:   MSEntra ID broker for authd
publisher: Canonical✓
store-url: https://snapcraft.io/authd-msentraid
license:   GPL-3.0
description: |
  This is the MS Entra ID broker snap for authd  to provide MS Entra ID OIDC based authentication on
  Ubuntu with authd.
services:
  authd-msentraid: simple, enabled, active
snap-id:      vS3oJLMss6lgWwoFcPqYDUA2HB20I1Dc
tracking:     0.x/edge
refresh-date: yesterday at 18:11 PST
channels:
  0.x/stable:    0.1+4fe9826.0f76acc 2024-10-02 (51) 18MB -
  0.x/candidate: ↑
  0.x/beta:      ↑
  0.x/edge:      0.1+a94565a.ed309a2 2024-11-12 (69) 18MB -
installed:       0.1+3c70d4e.048f86b            (67) 18MB -
```
* authd version: please run `/usr/libexec/authd version`
```
authd   0.3.6
```
* gnome shell version: please run `apt policy gnome-shell`
```
gnome-shell:
  Installed: 46.3.1-1ubuntu1~24.04.1authd2
  Candidate: 46.3.1-1ubuntu1~24.04.1authd2
  Version table:
 *** 46.3.1-1ubuntu1~24.04.1authd2 500
        500 https://ppa.launchpadcontent.net/ubuntu-enterprise-desktop/authd/ubuntu noble/main amd64 Packages
        100 /var/lib/dpkg/status
     46.0-0ubuntu6~24.04.5 500
        500 http://us.archive.ubuntu.com/ubuntu noble-updates/main amd64 Packages
     46.0-0ubuntu6~24.04.3 500
        500 http://security.ubuntu.com/ubuntu noble-security/main amd64 Packages
     46.0-0ubuntu5 500
        500 http://us.archive.ubuntu.com/ubuntu noble/main amd64 Packages
```
* Distribution: (**NAME** in `/etc/os-release`)
```
"Ubuntu"
```
* Distribution version: (**VERSION_ID** on `/etc/os-release`):
```
24.04
```

### Log files
Please redact/remove sensitive information:

#### Authd entries:

```raw
journalctl -u authd.service
```
N/A

####  MS Entra ID broker entries:

```raw
journalctl -u snap.authd-msentraid.authd-msentraid.service
```
N/A

### Application settings

Please redact/remove sensitive information:

####  Broker configuration:

```raw
cat /var/snap/authd-msentraid/current/broker.conf
```
```
[oidc]
issuer = https://login.microsoftonline.com/TENANT_ID/v2.0
client_id = CLIENT_ID

[users]
# The directory where the home directory will be created for new users.
# Existing users will keep their current directory.
# The user home directory will be created in the format of {home_base_dir}/{username}
home_base_dir = /home

# The username suffixes that are allowed to login via ssh without existing previously in the system.
# The suffixes must be separated by commas.
ssh_allowed_suffixes = @company.com
```

####  Broker authd configuration:

```raw
cat /etc/authd/brokers.d/msentraid.conf
```
```
# This section is used by authd to identify and communicate with the broker.
# It should not be edited.
[authd]
name = Microsoft Entra ID
brand_icon = /snap/authd-msentraid/current/broker_icon.png
dbus_name = com.ubuntu.authd.MSEntraID
dbus_object = /com/ubuntu/authd/MSEntraID
```

### Relevant information

N/A

### Double check your logs

- [X] I have redacted any sensitive information from the logs

**Comment by ajm370** [comment]
Hi @adombeck, thanks again for your time and detailed responses :)

`Those are indeed only available on user and device resources, so you can't use those to store a GID in a group resource. So it doesn't help with your goal of setting the same GID for the same group on different authd instances.`

I've thought on this more; I think we can make do with Extension Attributes as is, and would be very happy to test the patch if you and/or colleagues can put one together please.

Would there be an issue setting the GID of a user to be identical to their UID via authctl? I've noticed on all test Authd devices I've logged into that my GID and UID haven't matched (for my username@org.tld account, to clarify), and wasn't sure if that was simply due to randomisation or for a specific reason.

`That's a separate feature request, please file a GitHub issue for it.`

Sure, will do!

`authd creates the groups on the system and adds the users to them, so yes, they can be practically used. For example, if your user is a member of an Entra group "admin", a group of the same name will be created locally when you log in via authd. By default, /etc/sudoers allows members of a group named "admin" to use sudo. You could define other groups which should have sudo access in the sudoers file.

File permissions also work. Once the group was created locally (i.e. a user who is a member of that group logged in), you can make files owned by that group.

I don't understand what you mean with "excepting the login and SSH options present in the broker.conf file".`

That's fine, that all makes sense. Alongside with the previous part about 'linux-...' -> '...-linux' automatic group creation from Entra (to be a separate feature request), this will work for us. The login/SSH options are the part I understand definitely work, but require manual definition per-device/group of devices. I was asking about Entra group usage on Ubuntu authd-based systems so that we can manage permissions and trust via Entra policies & possibly other features like PIM, in a more automated way.

Many thanks again!

**Comment by adombeck** [comment]
> Would there be an issue setting the GID of a user to be identical to their UID via authctl? I've noticed on all test Authd devices I've logged into that my GID and UID haven't matched (for my [username@org.tld](mailto:username@org.tld) account, to clarify), and wasn't sure if that was simply due to randomisation or for a specific reason.

No, that will not be an issue. It currently doesn't match because we choose random IDs for users and groups. However, we actually plan to change that soon for user private groups, to use the same ID for the user and the user private group, as it should be according to https://wiki.debian.org/UserPrivateGroups#UPGs.

**Comment by ajm370** [comment]
Hi @adombeck, thanks again for putting the requested patch together.

Would it be possible to implement a similar function that pulls down on-premises AD custom extension attributes, **e.g. extensionAtrribute5** , for user who is logged in? I'm not sure if this could look something like the following request+response:

[https://learn.microsoft.com/en-us/graph/extensibility-overview?tabs=javascript#request](https://learn.microsoft.com/en-us/graph/extensibility-overview?tabs=javascript#request)

Cheers!

**Comment by adombeck** [comment]
Hi @ajm370!

> Would it be possible to implement a similar function that pulls down on-premises AD custom extension attributes, e.g. extensionAtrribute5 , for user who is logged in?

Maybe, but first I'd like to understand the use case that would solve. Since these extension attributes can't be used for groups, I don't think it makes sense to use these for synchronizing user and group IDs - I think the directory extensions make more sense for that. Can you explain why you want to use the on-premises AD extension attributes instead?
